doc/PROTOCOL-SEMANTICS.md

```diff
@@ -0,0 +1,221 @@
+the client application has not received the Trailers, then cancellation
+generally should win. No auxilary information is included in cancellation
+signals between the client and server. Server implementations may fail a Call
+and respond with Trailers while claiming to the server application that the Call
+was cancelled.
+
+There may be multiple servers that _could_ have received the Call (as is common
+for load balancing), but only a single server may process an individual Call.
+Calls are assumed to be non-idempotent and may not be "replayed" except for when
+gRPC is explicitly informed it is safe to do so.
+
+Calls are natively two independent streams (i.e., full duplex bidirectional) of
+Messages. The streams are each started with Headers. The end of the request
+stream is delimited by Half Close. The end of the response stream is delimited
+by Trailers. Messages may exist between the Headers and the end of the stream.
+Headers, Messages, Half Close, and Trailers are the units of communication and,
+absent the Call's termination, will be communicated to the remote without
+the need to send further units on the stream. However, see the optimizations
+permitted for unary Calls below.
+
+The two independent streams are each unidirectional and do not provide any
+information in the reverse direction other than Flow Control. Flow Control
+is a signal from the receiver to the sender to temporarily pause sending
+additional messages to avoid excessive buffering. Flow Control only applies to
+messages, but since streams are in-order Half Close or Trailers may be delayed
+waiting for message Flow Control in the same stream. No message receipt
+acknowledgement information is provided. However applications may use messages
+for such signals, as a response naturally acknowledges its request. Note in
+particular that there is no provision for the server application to not know
+whether the client application received a unary Call's response or a streaming
+Call's Status.
+
+Unary Calls may be optimized to be half-duplex and treat each stream as a single
+communication unit. That is, on the client a unary Call may be delayed from
+being sent until the Half Close is ready to be sent and on the server the
+response may be delayed until the Trailers is ready.
+
+Unary Calls that terminate with a Status Code of OK must contain a response
+message. Unary Calls that terminate with a Status Code other than OK do not need
+a response message, and at the implementation's discretion the response message
+may be discarded if present.
+
+## Status
+
+A Status contains a "code" and a "description". The Status Description is a
+human-readable Unicode string for developer debugging. The Status Code is a
+value from a pre-defined list of such codes. While Status Code is best
+communicated to users by its name, it commonly is treated as an integer
+internally, and so each code has a numeric value.
+
+The valid Status Codes are:
```

Comment:
This list feels like it should be in its own doc.  I thought we already had one, but the closest I see is https://github.com/grpc/grpc/blob/master/doc/statuscodes.md, which doesn't show this list directly.  How about either (a) adding this list to that doc or (b) creating a separate doc with this list, and then changing this section to be a link to that doc?